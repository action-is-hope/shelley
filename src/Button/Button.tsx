"use client";
import React, {
  Ref,
  ReactElement,
  useRef,
  forwardRef,
  ElementType,
} from "react";
import { useButton } from "react-aria";
import type { AriaButtonProps } from "@react-types/button";
import type {
  Accent,
  AlignPos,
  Volume,
  ButtonVariants,
} from "../typings/shared-types";
import type { MergeElementProps } from "../typings/utils";
import { useFocusRing } from "react-aria";
import { mergeRefs, mergeProps } from "@react-aria/utils";
import { st, classes } from "./button.st.css";

/**
 * Leveraging Adobes 'useButton' to enable use with their hooks.
 * N.B: Omitting 'elementType' as preference is to use existing 'as'.
 *
 * Adobe docs: https://react-spectrum.adobe.com/react-aria/useButton.html
 */

export interface ButtonCustomProps
  extends Omit<AriaButtonProps, "elementType"> {
  /** Define an Icon node, postion via #iconPos. */
  icon?: React.ReactNode;
  /** The position of the icon relative to the label. */
  iconPos?: AlignPos;
  /** Tone index, defines the color palette. */
  tone?: Accent;
  /** Variant index, defines the 'look'. */
  variant?: ButtonVariants;
  /** Defines how 'loud' the Button should be in term of it's size. */
  vol?: Volume;
  /** Applies width of 100%. */
  fullWidth?: boolean;
}

export type ButtonProps<P extends React.ElementType = "a" | "button" | any> = {
  /** Custom element to render such as an anchor "a" or a router "Link" component. */
  as?: P;
  // Dynamically apply element props types based on the input (P).
} & MergeElementProps<P, Omit<ButtonCustomProps, "as">>;

function Button<P extends React.ElementType = "button">(
  {
    as: As,
    children,
    className: classNameProp,
    icon,
    iconPos,
    fullWidth = false,
    tone = 1,
    variant = "primary",
    vol = 3,
    // Pull off known inputs for @react-aria -> useButton minus 'elementType'
    excludeFromTabOrder,
    isDisabled,
    onPress,
    onPressStart,
    onPressEnd,
    onPressChange,
    onPressUp,
    // autoFocus,
    onFocus,
    onBlur,
    onFocusChange,
    onKeyDown,
    onKeyUp,
    href,
    target,
    rel,
    ariaExpanded,
    ariaHaspopup,
    ariaControls,
    ariaPressed,
    type,
    id,
    ariaLabel,
    ariaLabelledby,
    ariaDescribedby,
    ariaDetails,
    ...rest
  }: ButtonProps<P>,
  ref: React.Ref<HTMLElement>
) {
  const localRef = useRef(null);

  const { buttonProps, isPressed } = useButton(
    {
      excludeFromTabOrder,
      isDisabled,
      onPress,
      onPressStart,
      onPressEnd,
      onPressChange,
      onPressUp,
      // autoFocus,
      onFocus,
      onBlur,
      onFocusChange,
      onKeyDown,
      onKeyUp,
      href,
      target,
      rel,
      "aria-expanded": ariaExpanded,
      "aria-haspopup": ariaHaspopup,
      "aria-controls": ariaControls,
      "aria-pressed": ariaPressed,
      type,
      id,
      "aria-label": ariaLabel,
      "aria-labelledby": ariaLabelledby,
      "aria-describedby": ariaDescribedby,
      "aria-details": ariaDetails,
      // Map 'as' to elementType for adobe-aria...
      elementType: (As as React.JSXElementConstructor<HTMLElement>) || "button",
    },
    localRef as React.RefObject<HTMLElement>
  );
  const { isFocusVisible, focusProps } = useFocusRing();

  const className = st(
    classes.root,
    {
      iconPos: icon ? iconPos : undefined,
      fullWidth,
      tone: tone !== false ? tone : undefined,
      variant: variant !== false ? variant : undefined,
      vol: vol !== false ? vol : undefined,
      isFocusVisible,
      isPressed,
      isDisabled,
    },
    classNameProp
  );

  const internal = (
    <>
      {icon && (
        <>
          {icon}
          {children && <span className={classes.divider}></span>}
        </>
      )}
      {children && <span className={classes.inner}>{children}</span>}
    </>
  );
  return React.createElement(
    As || "button",
    {
      ref: mergeRefs(localRef, ref),
      ...mergeProps(buttonProps, focusProps),
      className,
      ...rest,
    },
    internal
  );
}

/**
 * Buttons allow users to perform an action.
 */
// forwardRef doesn't support generic parameters -> cast to the correct type.
// https://stackoverflow.com/questions/58469229/react-with-typescript-generics-while-using-react-forwardref
const _Button = forwardRef(Button) as <P extends ElementType>(
  props: ButtonProps<P> & { ref?: Ref<HTMLElement> }
) => ReactElement;
/** required - see ButtonGroup */
_Button.toString = () => "ShelleyButton";
export { _Button as Button };
