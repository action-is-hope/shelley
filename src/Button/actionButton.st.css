@st-namespace "ActionButton";
@st-import ButtonBase, [
  --bg-color,
  --border-color,
  --text-color,
  --border-width,
  --border-radius] from "./buttonBase.st.css";

.root {
  -st-extends: ButtonBase;
  -st-states: isQuiet;
}
