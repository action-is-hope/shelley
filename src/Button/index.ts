export * from "./ButtonBase";
export * from "./Button";
export * from "./ActionButton";
export * from "./IconButton";
