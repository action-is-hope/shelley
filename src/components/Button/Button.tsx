import React from "react";
/* Adobe libs */
// Version dependancy issue: https://github.com/adobe/react-spectrum/issues/1388#issuecomment-781094658
import { useButton } from "react-aria";
import type { AriaButtonProps } from "@react-types/button";
/* Internal */
import type { Accent, AlignPos, Volume, ButtonVariants } from "../types";
import type { MergeElementProps } from "../utils";
/* Style API */
import { st, classes } from "./button.st.css";

/**
 * Leveraging Adobes 'useButton' to enable use with their hooks.
 * N.B: Omitting 'elementType' as preference is to use existing 'as'.
 *
 * Adobe docs: https://react-spectrum.adobe.com/react-aria/useButton.html
 */

export interface ButtonCustomProps
  extends Omit<AriaButtonProps, "elementType"> {
  /** Define an Icon node, postion via #iconPos. */
  icon?: React.ReactNode;
  /** The position of the icon relative to the label. */
  iconPos?: AlignPos;
  /** Tone index, defines the color palette. */
  tone?: Accent;
  /** Variant index, defines the 'look'. */
  variant?: ButtonVariants;
  /** Defines how 'loud' the Button should be in term of it's size. */
  vol?: Volume;
  /** Applies width of 100%. */
  fullWidth?: boolean;
}

export type ButtonProps<P extends React.ElementType = "button"> = {
  /** Custom element to render such as an anchor "a" or a router "Link" component. */
  as?: P;
  // Dynamically apply element props types based on the input (P).
} & MergeElementProps<P, Omit<ButtonCustomProps, "as">>;

function ButtonBase<T extends React.ElementType = "button">(
  {
    as: As,
    children,
    className: classNameProp,
    icon,
    iconPos = "end",
    fullWidth = false,
    tone = 1,
    variant = "primary",
    vol = 3,
    // Pull off known inputs for @react-aria -> useButton minus 'elementType'
    isDisabled,
    onPress,
    onPressStart,
    onPressEnd,
    onPressChange,
    onPressUp,
    autoFocus,
    onFocus,
    onBlur,
    onFocusChange,
    onKeyDown,
    onKeyUp,
    href,
    target,
    rel,
    ariaExpanded,
    ariaHaspopup,
    ariaControls,
    ariaPressed,
    type,
    id,
    ariaLabel,
    ariaLabelledby,
    ariaDescribedby,
    ariaDetails,
    ...rest
  }: ButtonProps<T>,
  ref: React.Ref<HTMLElement>
) {
  const { buttonProps, isPressed } = useButton(
    {
      isDisabled,
      onPress,
      onPressStart,
      onPressEnd,
      onPressChange,
      onPressUp,
      autoFocus,
      onFocus,
      onBlur,
      onFocusChange,
      onKeyDown,
      onKeyUp,
      href,
      target,
      rel,
      "aria-expanded": ariaExpanded,
      "aria-haspopup": ariaHaspopup,
      "aria-controls": ariaControls,
      "aria-pressed": ariaPressed,
      type,
      id,
      "aria-label": ariaLabel,
      "aria-labelledby": ariaLabelledby,
      "aria-describedby": ariaDescribedby,
      "aria-details": ariaDetails,
      // Map 'as' to elementType for adobe-aria...
      // ...rest,
      elementType: (As as React.JSXElementConstructor<HTMLElement>) || "button",
    },
    ref as React.RefObject<HTMLElement>
  );

  const className = st(
    classes.root,
    {
      iconPos: icon ? iconPos : undefined,
      fullWidth,
      tone,
      variant: variant !== false ? variant : undefined,
      vol: vol !== false ? vol : undefined,
      isPressed,
      isDisabled,
    },
    classNameProp
  );
  const internal = (
    <>
      {icon && (
        <>
          {icon}
          {children && <span className={classes.divider}></span>}
        </>
      )}
      {children && <span className={classes.inner}>{children}</span>}
    </>
  );
  return React.createElement(
    As || "button",
    {
      ref,
      ...buttonProps,
      className,
      ...rest,
    },
    internal
  );
}

// See:
// https://github.com/adobe/react-spectrum/blob/main/packages/%40react-spectrum/listbox/src/ListBox.tsx

// const Button = React.forwardRef(ButtonBase);
const Button = React.forwardRef(ButtonBase) as typeof ButtonBase;

// Button.displayName = "Button";

export default Button;
