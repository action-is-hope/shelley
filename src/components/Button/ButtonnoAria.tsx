// import * as React from "react";
import type { ElementType, ReactNode, Ref } from "react";
import { createElement, forwardRef } from "react";
import type { Accent, AlignPos, Volume, ButtonVariants } from "../types";
import type { MergeElementProps } from "../utils";
/* = Style API. */
import { st, classes } from "./button.st.css";

/**
 * Props
 */

export interface ButtonBaseProps {
  /**
   * Define an Icon node; position the icon via #iconPos.
   **/
  icon?: ReactNode;
  /**
   * The position of the icon relative to the label.
   **/
  iconPos?: AlignPos;
  /**
   * Tone index, changes color of button if required.
   **/
  tone?: Accent;
  /**
   * Variant index, defines the 'look'.
   **/
  variant?: ButtonVariants;
  /**
   * Defines how 'loud' the Button should be in term of it's size.
   **/
  vol?: Volume;
  /**
   * Applies width of 100%.
   **/
  fullWidth?: boolean;
}

export type ButtonProps<P extends ElementType = "button"> = {
  /**
   * Custom element to render such as an anchor "a" or a router "Link" component.
   **/
  as?: P;
} & MergeElementProps<P, ButtonBaseProps>;

export function ButtonBase<T extends ElementType = "button">(
  {
    children,
    className: classNameProp,
    as,
    icon,
    iconPos = "end",
    fullWidth = false,
    tone = 1,
    variant = "primary",
    // onPress,
    // onPressStart,
    vol = 3,
    role,
    ...rest
  }: ButtonProps<T>,
  ref: Ref<T>
) {
  const className = st(
    classes.root,
    {
      iconPos: icon ? iconPos : undefined,
      fullWidth,
      tone,
      variant,
      vol,
    },
    classNameProp
  );
  const internal = (
    <>
      {icon && (
        <>
          {icon}
          {children && <span className={classes.divider}></span>}
        </>
      )}
      {children && <span className={classes.inner}>{children}</span>}
    </>
  );
  return createElement(
    as || "button",
    {
      ref,
      ...rest,
      // role: as === "a" ? undefined : role,
      role,
      className,
    },
    internal
  );
}

// const Button = forwardRef(ButtonBase) as typeof ButtonBase;
const Button = forwardRef(ButtonBase);

Button.displayName = "Button";

export default Button;
