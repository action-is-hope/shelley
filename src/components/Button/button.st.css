/* button.st.css */
.root {
  -st-states: iconPos(enum(top, end, bottom, start));
  display: inline-flex;
  justify-content: space-between;
}

/*==== Input Positions ====*/

/* top */
.root:iconPos(top) {
  flex-direction: column-reverse;
}

/* end / right */
.root:iconPos(end) {
  flex-direction: row-reverse;
}

/* bottom */
.root:iconPos(bottom) {
  flex-direction: column;
}

/* start / left */
.root:iconPos(start) {
  flex-direction: row;
}

/* The 'inner' button. */
.inner {}

/* The divider which is rendered with an icon only. */
.divider {}

/* The infotip element for inline tooltip style on focus/hover. */
.tip {}

/* Tones */

.tone1 {}

.tone2 {}

.tone3 {}

.tone4 {}

.tone5 {}

.tone6 {}


/* Variants */

.variant1 {}

.variant2 {}

.variant3 {}

.variant4 {}

.variant5 {}

.variant6 {}


/* Volume */

.vol1 {}

.vol2 {}

.vol3 {}

.vol4 {}

.vol5 {}

.vol6 {}
