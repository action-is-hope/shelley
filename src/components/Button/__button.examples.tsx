/* buttonExamples.tsx */
import React from "react";
/* Example Component */
import Button from "./Button";
/* Supporting Components */
import Icon from "../../components/Icon/Icon";
import PropsDemo from "../../components_site/PropsDemo/PropsDemo";
import CodeSample from "../../components_site/CodeSample/CodeSample";
import { classes as grid } from "../../styles/default/grid.st.css";
export const meta = {
  name: "HTMLInputElement"
};

export const QuickRef = () => (
  <CodeSample
    className={grid.mid}
  >{`import Button from "@action-is-hope/shelley";
\n<Button>Climate fight club</Button>
/* One rule: We don't fight, we act on climate change with intent. */`}</CodeSample>
);

// Props Demo - Examples below.
export const ComponentDemo = () => {
  const [labelDemoProps, setLabelDemoProps]: any = React.useState([
    {
      name: "children",
      label: "children",
      type: "text",
      value: "Earth Solutionists required"
    },
    {
      name: "tone",
      label: "tone",
      type: "number",
      min: 0,
      max: 6,
      value: 1
    },
    {
      name: "variant",
      label: "variant",
      type: "number",
      min: 0,
      max: 6,
      value: 1
    },
    {
      name: "volume",
      label: "volume",
      type: "number",
      min: 0,
      max: 6,
      value: 2
    },
    {
      name: "icon",
      label: "icon",
      type: "select",
      value: "No",
      options: ["Yes", "No"]
    },
    {
      name: "tip",
      label: "tip",
      type: "text",
      value: "@todo - unsure on this..."
    }
  ]);
  return (
    <PropsDemo
      demoProps={labelDemoProps}
      componentName="Button"
      setDemoProps={setLabelDemoProps}
      tsExtends="HTMLButtonElement"
      renderExample={
        <CodeSample>{`import { Button, Icon } from "@action-is-hope/shelley";\n\n<Button ${
          labelDemoProps[1].value > 0
            ? `\n  tone={${labelDemoProps[1].value}}`
            : `\n  tone={false}`
        } ${
          labelDemoProps[2].value > 0
            ? `\n  variant={${labelDemoProps[2].value}}`
            : `\n  variant={false}`
        } ${
          labelDemoProps[3].value > 0
            ? `\n  vol={${labelDemoProps[3].value}}`
            : `\n  vol={false}`
        } ${
          labelDemoProps[4].value === "Yes"
            ? `\n  icon={<Icon><path d="M14 7h-5v-5h-2v5h-5v2h5v5h2v-5h5v-2z"></path></Icon>}`
            : ``
        }  \n>\n ${labelDemoProps[0].value}\n</Button>`}</CodeSample>
      }
    >
      <Button
        tone={labelDemoProps[1].value === 0 ? false : labelDemoProps[1].value}
        icon={
          labelDemoProps[4].value === "No" ? (
            false
          ) : (
            <Icon alt={labelDemoProps[1].value}>
              {/* crack open an svg to find the path and shove it in. */}
              <path d="M14 7h-5v-5h-2v5h-5v2h5v5h2v-5h5v-2z"></path>
              {/* <path d="M16 14h-16v-12h16v12zM1 13h14v-10h-14v10z"></path> */}
              {/* <path d="M2 10v2h12v-1c0 0 0.2-1.7-2-2-1.9-0.3-2.2 0.6-3.8 0.6-1.1 0-0.9-1.6-3.2-1.6-1.7 0-3 2-3 2z"></path> */}
              {/* <path d="M13 6c0 1.105-0.895 2-2 2s-2-0.895-2-2c0-1.105 0.895-2 2-2s2 0.895 2 2z"></path> */}
            </Icon>
          )
        }
        variant={
          labelDemoProps[2].value === 0 ? false : labelDemoProps[2].value
        }
        vol={labelDemoProps[3].value === 0 ? false : labelDemoProps[3].value}
        // tip="hi"
      >
        {labelDemoProps[0].value}
      </Button>
      {/* <Button
        tone={labelDemoProps[1].value === 0 ? false : labelDemoProps[1].value}
        icon={
          <Icon>
            <path d="M14 7h-5v-5h-2v5h-5v2h5v5h2v-5h5v-2z"></path>
          </Icon>
        }
        variant={
          labelDemoProps[2].value === 0 ? false : labelDemoProps[2].value
        }
        vol={labelDemoProps[3].value === 0 ? false : labelDemoProps[3].value}
      >
        &nbsp;
      </Button> */}
    </PropsDemo>
  );
};

export const ComponentHTML = () => (
  <CodeSample>{`<button class="button-root button-tone1 button-variant1 button-vol2">
  <span class="button-inner">Earth Guardians</span>
</button>`}</CodeSample>
);

export const ComponentCSS = () => (
  <CodeSample fixedHeight>{`/** button.st.css */

/* Root */
Button {}

/* Parts */

/* 'inner' button, use for vertical spacing to give a potential Icon more room */
Button::inner {}

/* The divider which is rendered with an icon only. */
Button::divider {}

/* A basic CSS tooltip. */
Button::tip {}


/* Accent colours */
Button.tone1 {}
Button.tone2 {}
Button.tone3 {}
Button.tone4 {}
Button.tone5 {}
Button.tone6 {}

/* Volumes */
Button.vol1 {}
Button.vol2 {}
Button.vol3 {}
Button.vol4 {}
Button.vol5 {}
Button.vol6 {}

/* Variants */
Button.variant1 {}
Button.variant2 {}
Button.variant3 {}
Button.variant4 {}
Button.variant5 {}
Button.variant6 {}`}</CodeSample>
);

// export const Example1 = () => <Demo></Demo>;
// Add some tests to check refs and tings are spread properly.
// const inputEl = React.useRef();
// const test = React.createRef<HTMLButtonElement>();
// const onButtonClick = () => {
//   const node = test.current;
//   // `current` points to the mounted text input element
//   node && node.focus();
// };
