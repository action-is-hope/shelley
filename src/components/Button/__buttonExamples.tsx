/* buttonExamples.tsx */
import React from "react";
// Example Component
import Button from "./Button";
// Supporting Components
import PropsDemo from "../../components_site/PropsDemo/PropsDemo";
import CodeSample from "../../components_site/CodeSample/CodeSample";

export const meta = {
  name: "Button"
};

export const QuickRef = () => (
  <CodeSample>{`import Button from "@action-is-hope/shelley";
\n<Button>Climate action now</Button>`}</CodeSample>
);

// Props Demo - Examples below.
export const ComponentDemo = () => {
  const [labelDemoProps, setLabelDemoProps]: any = React.useState([
    {
      name: "children",
      label: "children",
      type: "text",
      value: "Yes to climate action"
    },
    {
      name: "color",
      label: "color",
      type: "number",
      min: 0,
      max: 6,
      value: 1
    },
    {
      name: "variant",
      label: "variant",
      type: "number",
      min: 0,
      max: 6,
      value: 1
    },
    {
      name: "volume",
      label: "volume",
      type: "number",
      min: 0,
      max: 6,
      value: 2
    },
    {
      name: "cite",
      label: "cite",
      type: "text",
      value: "Donella Meadows, environmental scientist, 1941-2001 "
    },
    {
      name: "citeUrl",
      label: "citeUrl",
      type: "text",
      value: "https://en.wikipedia.org/wiki/Donella_Meadows"
    }
  ]);
  return (
    <PropsDemo
      demoProps={labelDemoProps}
      componentName="Button"
      setDemoProps={setLabelDemoProps}
      tsExtends="HTMLButtonElement"
      renderExample={
        <CodeSample>{`import { Button } from "@action-is-hope/shelley";\n\n<Button ${
          labelDemoProps[1].value > 0
            ? `\n  color={${labelDemoProps[1].value}}`
            : `\n  color={false}`
        } ${
          labelDemoProps[2].value > 0
            ? `\n  variant={${labelDemoProps[2].value}}`
            : `\n  variant={false}`
        } ${
          labelDemoProps[3].value > 0
            ? `\n  volume={${labelDemoProps[3].value}}`
            : `\n  volume={false}`
        } ${
          labelDemoProps[1].value
            ? `\n  citeUrl="${labelDemoProps[3].value}"`
            : ``
        }  \n>\n ${labelDemoProps[0].value}\n</Button>`}</CodeSample>
      }
    >
      <Button
        color={labelDemoProps[1].value === 0 ? false : labelDemoProps[1].value}
        variant={
          labelDemoProps[2].value === 0 ? false : labelDemoProps[2].value
        }
        vol={labelDemoProps[3].value === 0 ? false : labelDemoProps[3].value}
      >
        {labelDemoProps[0].value}
      </Button>
    </PropsDemo>
  );
};

export const ComponentHTML = () => (
  <CodeSample>{`<button class="button--root button--accent1 button--variant1 button--vol2">
  <span class="button--inner">Yes to climate action</span>
</button>`}</CodeSample>
);

export const ComponentCSS = () => (
  <CodeSample fixedHeight>{`/** button.st.css */

/* Root */
Button {}

/* Parts */

/* 'inner' button, use for vertical spacing to give a potential Icon more room */
Button::inner {}

/* The divider which is rendered with an icon only. */
Button::divider {}

/* A basic CSS tooltip. */
Button::tip {}


/* Accent colours */
Button.accent1 {}
Button.accent2 {}
Button.accent3 {}
Button.accent4 {}
Button.accent5 {}
Button.accent6 {}

/* Volumes */
Button.vol1 {}
Button.vol2 {}
Button.vol3 {}
Button.vol4 {}
Button.vol5 {}
Button.vol6 {}

/* Variants */
Button.variant1 {}
Button.variant2 {}
Button.variant3 {}
Button.variant4 {}
Button.variant5 {}
Button.variant6 {}`}</CodeSample>
);

// export const Example1 = () => <Demo></Demo>;
// Add some tests to check refs and tings are spread properly.
// const inputEl = React.useRef();
// const test = React.createRef<HTMLButtonElement>();
// const onButtonClick = () => {
//   const node = test.current;
//   // `current` points to the mounted text input element
//   node && node.focus();
// };
