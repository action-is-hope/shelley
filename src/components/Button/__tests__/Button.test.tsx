import React from "react";
import Button from "../Button";
import { render } from "@testing-library/react";

const buttonRef = React.createRef<HTMLButtonElement>();
// const onButtonClick = () => {
//   const node = buttonRef.current;
//   // `current` points to the mounted text input element
//   node && node.focus();
// };

// You have to write data-testid
const ButtonTest = () => (
  <Button data-testid="button-text" ref={buttonRef}>
    Gatsby is awesome!
  </Button>
);
test("Displays the correct button text.", () => {
  const { getByTestId } = render(<ButtonTest />);
  // Assertion
  expect(getByTestId("button-text")).toHaveTextContent("Gatsby is awesome!");
  // --> Test will pass
});
