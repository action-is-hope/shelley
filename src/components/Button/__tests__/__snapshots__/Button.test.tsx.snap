// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button includes #icon via prop positioned via #iconPos divider and children 1`] = `
<button
  className="Button__Test__root Button__Test---iconPos-5-start Button__Test---tone-1-1 Button__Test---variant-7-primary Button__Test---vol-1-3"
  data-testid="button-data-testid"
  onBlur={null}
  onClick={[Function]}
  onDragStart={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  type="button"
>
  <svg
    aria-hidden={true}
    className="Icon__Test__root"
    focusable="false"
    viewBox="0 0 16 16"
  >
    <path
      d="M14 7h-5v-5h-2v5h-5v2h5v5h2v-5h5v-2z"
    />
  </svg>
  <span
    className="Button__Test__divider"
  />
  <span
    className="Button__Test__inner"
  >
    I love big buttons and I cannot lie...
  </span>
</button>
`;

exports[`Button includes #icon, no children, visually hidden label text provided via Icon alt prop 1`] = `
<button
  className="Button__Test__root Button__Test---iconPos-3-end Button__Test---tone-1-1 Button__Test---variant-9-secondary Button__Test---vol-1-3"
  data-testid="button-data-testid"
  onBlur={null}
  onClick={[Function]}
  onDragStart={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  type="button"
>
  <svg
    aria-hidden={true}
    className="Icon__Test__root"
    focusable="false"
    viewBox="0 0 16 16"
  >
    <path
      d="M14 7h-5v-5h-2v5h-5v2h5v5h2v-5h5v-2z"
    />
  </svg>
  <span
    className="visuallyHidden__Test__root"
  >
    I love big buttons and I cannot lie...
  </span>
</button>
`;

exports[`Button renders as a basic anchor link via #href prop 1`] = `
<a
  className="Button__Test__root Button__Test---tone-1-1 Button__Test---variant-7-primary Button__Test---vol-1-3"
  data-testid="button-data-testid"
  href="http://shelley.earth"
  onBlur={null}
  onClick={[Function]}
  onDragStart={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    className="Button__Test__inner"
  >
    I love big buttons and I cannot lie...
  </span>
</a>
`;

exports[`Button renders as a basic button with custom #className 1`] = `
<button
  className="Button__Test__root Button__Test---tone-1-1 Button__Test---variant-7-primary Button__Test---vol-1-3 for-which-nobody-can-deny"
  data-testid="button-data-testid"
  onBlur={null}
  onClick={[Function]}
  onDragStart={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  type="button"
>
  <span
    className="Button__Test__inner"
  >
    I love big buttons and I cannot lie...
  </span>
</button>
`;

exports[`Button renders as third party (react-router) anchor provided via #as and #to props 1`] = `
<a
  className="Button__Test__root Button__Test---tone-1-1 Button__Test---variant-7-primary Button__Test---vol-1-3"
  data-testid="button-data-testid"
  href="/internal-link"
  onBlur={null}
  onClick={[Function]}
  onDragStart={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    className="Button__Test__inner"
  >
    I love big buttons and I cannot lie...
  </span>
</a>
`;

exports[`Button renders with stylable state prop classes, #tone, #variant, #vol and #fullwidth 1`] = `
<button
  className="Button__Test__root Button__Test--fullWidth Button__Test---tone-1-6 Button__Test---variant-7-primary Button__Test---vol-1-6"
  data-testid="button-data-testid"
  onBlur={null}
  onClick={[Function]}
  onDragStart={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  type="button"
>
  <span
    className="Button__Test__inner"
  >
    I love big buttons and I cannot lie...
  </span>
</button>
`;
